fsrs/src/test/kotlin/com/inout/fsrs/FSRSTest.kt

```diff
@@ -0,0 +1,338 @@
+                Calendar.getInstance().apply {
+                    set(2024, 0, 1, 0, 0, 0)
+                }.time
+
+            // when
+            val firstLogs = fsrs.repeat(card, firstDue).logs
+            val secondDue = firstLogs[Grade.Easy]!!.card.due
+            val secondLogs = fsrs.repeat(firstLogs[Grade.Easy]!!.card, secondDue).logs
+
+            // then
+            val againLog = secondLogs[Grade.Again]!!
+            assertEquals(State.Relearning, againLog.card.state)
+            assertEquals(secondDue.addMinutes(5), againLog.card.due)
+            assertEquals(11, againLog.card.elapsedDays)
+            assertEquals(0, againLog.card.scheduledDays)
+            assertEquals(2, againLog.card.reps)
+            assertEquals(1, againLog.card.lapses)
+            assertEquals(secondDue, againLog.card.lastReview)
+
+            val hardLog = secondLogs[Grade.Hard]!!
+            assertEquals(State.Review, hardLog.card.state)
+            assertEquals(secondDue.addDays(18), hardLog.card.due)
+            assertEquals(11, hardLog.card.elapsedDays)
+            assertEquals(18, hardLog.card.scheduledDays)
+            assertEquals(2, hardLog.card.reps)
+            assertEquals(0, hardLog.card.lapses)
+            assertEquals(secondDue, hardLog.card.lastReview)
+
+            val goodLog = secondLogs[Grade.Good]!!
+            assertEquals(State.Review, goodLog.card.state)
+            assertEquals(secondDue.addDays(42), goodLog.card.due)
+            assertEquals(11, goodLog.card.elapsedDays)
+            assertEquals(42, goodLog.card.scheduledDays)
+            assertEquals(2, goodLog.card.reps)
+            assertEquals(0, goodLog.card.lapses)
+            assertEquals(secondDue, goodLog.card.lastReview)
+
+            val easyLog = secondLogs[Grade.Easy]!!
+            assertEquals(State.Review, easyLog.card.state)
+            assertEquals(secondDue.addDays(98), easyLog.card.due)
+            assertEquals(11, easyLog.card.elapsedDays)
+            assertEquals(98, easyLog.card.scheduledDays)
+            assertEquals(2, easyLog.card.reps)
+            assertEquals(0, easyLog.card.lapses)
+            assertEquals(secondDue, easyLog.card.lastReview)
+        }
+    }
+
+    @Nested
+    inner class Rollback {
+        @Test
+        fun `rollback should undo the repeat`() {
+            // given
+            val fsrs = createFSRS()
+            val card = createCard()
+            val cardCopied = card.copy()
+            val due =
+                Calendar.getInstance().apply {
+                    set(2024, 0, 1, 0, 0, 0)
+                }.time
+            val firstRepeat = fsrs.repeat(card, due)
+
+            // when & then
+            listOf(Grade.Again, Grade.Hard, Grade.Good, Grade.Easy).forEach {
+                val recordLogItem = firstRepeat.logs[it]!!
+                val rollbackCard = fsrs.rollback(recordLogItem.card, recordLogItem.log)
+                assertEquals(cardCopied, rollbackCard)
```

Comment:
<div>

<strong>Bito Code Review Agent Run</strong> <a href=https://github.com/in-out-team/api-server/pull/10#issuecomment-2358174366>#8da850</a><span> - 09/18/2024, 11:15 am</span>
<br>

<div><small>🟠</small> Medium importance</div>

<div id="suggestion">
<div id="issue">
<b>Issue:</b> The rollback test case does not verify individual properties of the card after rollback
</div>
<div id="fix">
<b>Fix:</b> Enhance the rollback test to verify individual properties of the card after rollback
</div>
<details>
<summary>
<b>Code suggestion</b>
</summary>
<div id="code">

```
 @@ -306,7 +306,15 @@
          listOf(Grade.Again, Grade.Hard, Grade.Good, Grade.Easy).forEach {
              val recordLogItem = firstRepeat.logs[it]!!
              val rollbackCard = fsrs.rollback(recordLogItem.card, recordLogItem.log)
 -            assertEquals(cardCopied, rollbackCard)
 +            assertEquals(cardCopied.state, rollbackCard.state, "State should match after rollback")
 +            assertEquals(cardCopied.due, rollbackCard.due, "Due date should match after rollback")
 +            assertEquals(cardCopied.stability, rollbackCard.stability, "Stability should match after rollback")
 +            assertEquals(cardCopied.difficulty, rollbackCard.difficulty, "Difficulty should match after rollback")
 +            assertEquals(cardCopied.elapsedDays, rollbackCard.elapsedDays, "Elapsed days should match after rollback")
 +            assertEquals(cardCopied.scheduledDays, rollbackCard.scheduledDays, "Scheduled days should match after rollback")
 +            assertEquals(cardCopied.reps, rollbackCard.reps, "Reps should match after rollback")
 +            assertEquals(cardCopied.lapses, rollbackCard.lapses, "Lapses should match after rollback")
 +            assertEquals(cardCopied.lastReview, rollbackCard.lastReview, "Last review should match after rollback")
          }
      }
  }
```
</div>
</details>
</div>
<br>


</div>

Is this a valid issue, or was it incorrectly flagged by the Agent?
- [ ] it was incorrectly flagged